core/src/main/java/org/apache/accumulo/core/rpc/clients/TServerClient.java

```diff
@@ -57,44 +60,54 @@ default Pair<String,C> getTabletServerConnection(Logger LOG, ThriftClientTypes<C
       ClientContext context, boolean preferCachedConnections, AtomicBoolean warned)
       throws TTransportException {
     checkArgument(context != null, "context is null");
-    long rpcTimeout = context.getClientTimeoutInMillis();
-    // create list of servers
-    ArrayList<ThriftTransportKey> servers = new ArrayList<>();
+    final long rpcTimeout = context.getClientTimeoutInMillis();
 
-    // add tservers
-    ZooCache zc = context.getZooCache();
-    for (String tserver : zc.getChildren(context.getZooKeeperRoot() + Constants.ZTSERVERS)) {
-      var zLocPath =
-          ServiceLock.path(context.getZooKeeperRoot() + Constants.ZTSERVERS + "/" + tserver);
-      byte[] data = zc.getLockData(zLocPath);
-      if (data != null) {
-        String strData = new String(data, UTF_8);
-        if (!strData.equals("manager")) {
-          servers.add(new ThriftTransportKey(
-              new ServerServices(strData).getAddress(Service.TSERV_CLIENT), rpcTimeout, context));
+    final ZooCache zc = context.getZooCache();
+    final List<String> tservers = new ArrayList<>();
+    final AtomicBoolean warnedAboutTServersBeingDown = new AtomicBoolean(false);
+
+    for (int retries = 0; retries < 10; retries++) {
```

Comment:
The existing code did not retry and poking around it seems that code that calls this method does retry at higher levels. It would probably be better to not retry here and defer to the calling code.  Although I don't fully understand the overall behavior of the existing code when the set of tservers in ZK is empty, sitll trying to understand that.  Not exaclty sure how the existing code would retry. It seems like the addition of throwing TTransportException in this PR will help with the higher level code that retries, but not sure how it worked before these changes.

Reply:
This is [one example](https://github.com/apache/accumulo/blob/c72bf8c00cb196d048d1aa35077327c55d6e58e2/core/src/main/java/org/apache/accumulo/core/rpc/clients/TServerClient.java#L100-L125) of code I was looking at where retry happens at a higher level that calls this method indirectly.

Reply:
I removed the retry and while loop in aecd790